module4-authentication-and-security/securing-backend-app/README.md

```diff
@@ -0,0 +1,79 @@
+| 2   | Philips Blender      | 2       | https://url.com/313   | $90   | false  |
+| 3   | AMD Ryzen 9          | 1       | https://amzn.com/585  | $599  | true   |
+| 4   | Tefal Iceforce knife | 2       | https://amzn.com/1311 | $20   | false  |
+
+Now, suppose another user comes to your app, how will your app decides what to show them? When they land on the homepage, they would see my items because the code can't decide what to show. This is the first security risk, the app needs somehow to relate the data to a specific user, so it only query data that belongs to the requesting user.
+
+That's why almost everywhere online, you need to register or login to an account in order to use an application. So tha data can be related to you and you only.
+
+## Let's add some security
+
+To secure our wishlist above, we can add a new column to the lists table to attach it to an identity (a user). We can name that column `user_email` for example, and use this query to get the user wishlist:
+
+```SQL
+SELECT `id`, `name`, `user_email` from `lists`
+INNER JOIN `items`
+ON `items`.`list_id` = `lists`.`id`
+WHERE `lists`.`user_email` = "test@domain.com"
+```
+
+Thus, when a user comes to your app, you can ask them first about their email address, and then query the database to get thier lists. But that is also insufficient. A malicious user can peep on other users wishlists if he/she knew thier emails, or he/she check different email combinations to see if it lands a hit, aka, **Brute Force**.
+
+To make this a bit harder to brute force, you can create a token system. Tokens are random strings of charcters, usually wih arbitrary length, that is intended to be hard to brute force (guess). A token can look something like this: `1FsahNjuio31OHsOallsn1gHH7`
+
+You can change the column `user_email` to `token` and everytime a user comes in, you can ask them for the token, or give them the opportunity to generate a token. If they are new users, they can generate a new token, and save it, then start creating thier lists. When they visit again, they enter that token to get thier lists back.
+
+Bingo! You've just created your first, very acceptable authentication layer! You've secured your app and **slimmed the chances** of brute forcing user data! Many apps today use this pattern to secure data, however it is not always acceptable. That's because, if a user lost thier token, it is next to impossible (especially if the token was generated by a cryptograpic random generators with reasonable length) to to gain their access back to their lists using today's technologies.
+
+## Why we need passwords
+
+The token system developed above is very robust, but it isn't always acceptable security measure (at least not alone). That's because, if a user lost thier token, it is next to impossible (especially if the token was generated by a cryptograpic random generators with reasonable length) to to gain their access back to their lists using today's technologies.
```

Comment:
"especially if the token was generated by a cryptograpic random generators with reasonable length"

this sentence will most likely mean nothing to the students except to confuse them, so it should be either deleted or further clarified

Reply:
Yes, I am intending to add an example to it